test/rnn_seq_api.cpp

```diff
@@ -0,0 +1,176 @@
+    {
+        if(this->useDropout == 1 && (this->hiddenSize == 1 || this->batchSize == 1))
+            return false;
+
+        if(this->inputMode == 1 && this->hiddenSize != this->inVecLen)
+            return false;
+
+        if((this->rnnMode != 2) && (!this->nocx || !this->nocy))
+            return false;
+
+        if(this->seqLenArray.size() > this->batchSize)
+            return false;
+
+        if(!this->seqLenArray.empty())
+        {
+            if(this->seqLength <
+               *std::max_element(this->seqLenArray.begin(), this->seqLenArray.end()))
+                return false;
+
+            if(this->io_layout == 1)
+            {
+                return std::is_sorted(
+                    this->seqLenArray.begin(), this->seqLenArray.end(), std::greater<int>());
+            }
+        }
+        return true;
+    }
+
+    void run()
+    {
+        if(!this->full_set || (is_correct_params() && !is_skip_comb()))
+            rnn_seq_api_test_driver<T>::run();
+        else
+        {
+            if(this->verbose)
+                std::cout << "Incompatible argument combination, test skipped: "
+                          << this->get_command_args() << std::endl;
+        }
+    }
+};
+
+template <class T>
+struct lstm_MS_solver : rnn_seq_driver<T>
+{
+    lstm_MS_solver() : rnn_seq_driver<T>(true)
+    {
+        std::vector<int> modes(2, 0);
+        modes[1] = 1;
+
+        this->add(this->inVecLen, "vector-len", this->generate_data(std::vector<int>{1, 7}, 7));
+        this->add(this->hiddenSize, "hidden-size", this->generate_data({13, 1}, 13));
+        this->add(this->useDropout, "use-dropout", this->generate_data({0}));
+        this->add(this->numLayers, "num-layers", this->generate_data({3}));
+        this->add(this->inputMode, "in-mode", this->generate_data({0}));
+        this->add(this->biasMode, "bias-mode", this->generate_data(modes));
+        this->add(this->dirMode, "dir-mode", this->generate_data({0}));
+        this->add(this->rnnMode, "rnn-mode", this->generate_data({2}));
+        this->add(this->algoMode, "algo-mode", this->generate_data({0}));
+
+        this->add(this->io_layout, "io_layout", this->generate_data({2}, 2));
+        this->add(this->batchSize, "batch-size", this->generate_data({1, 6}, 6));
+        this->add(this->seqLength, "seq-len", this->generate_data({38}));
+        this->add(this->seqLenArray,
+                  "seqLen-batch",
+                  this->generate_data({
+                      {34, 3, 2, 1},
+                      {1, 15, 34, 15, 34, 1},
+                      {},
+                  }));
+
+        this->add(this->nohx, "nohx", this->generate_data(modes));
+        this->add(this->nocx, "nocx", this->generate_data(modes));
+        this->add(this->nohy, "nohy", this->generate_data(modes));
+        this->add(this->nocy, "nocy", this->generate_data(modes));
+    }
+
+    void run()
+    {
+        if(this->nohx && this->biasMode == 1)
+            return;
```

Comment:
[Q] Is it fine that it returns silently?

Reply:
The current implementation is incorrect for this case. The library and tests think that everything is fine, but it is not, it will be fixed later so not to increase this PR. The old tests never used nohx=true.

Reply:
[resolved] @shurale-nkn Good, thanks.